frameworks/MLPlan/exec.py

```diff
@@ -0,0 +1,101 @@
+import logging
+import math
+import os
+import json
+
+from frameworks.shared.callee import call_run, result, output_subdir, utils
+
+log = logging.getLogger(__name__)
+
+def run(dataset, config):
+    log.info("\n**** ML-Plans ****\n")
+
+    is_classification = config.type == 'classification'
+    
+    # Mapping of benchmark metrics to Weka metrics
+    metrics_mapping = dict(
+        acc='ERRORRATE',
+        auc='AUC',
+        logloss='LOGLOSS',
+        f1='F1',
+        r2='R2',
+        rmse='ROOT_MEAN_SQUARED_ERROR',
+        mse='MEAN_SQUARED_ERROR',
+        rmsle='ROOT_MEAN_SQUARED_LOGARITHM_ERROR',
+        mae='MEAN_ABSOLUTE_ERROR'
+    )
+
+    metric = metrics_mapping[config.metric] if config.metric in metrics_mapping else None
+    if metric is None:
+	    raise ValueError('Performance metric {} is not supported.'.format(config.metric))
+    
+    train_file = dataset.train.path
+    test_file = dataset.test.path
+
+    training_params = {k: v for k, v in config.framework_params.items() if not k.startswith('_')}
+    backend = config.framework_params.get('_backend', 'weka')
+
+    if backend == "weka":
+        mem_limit = str(max(config.max_mem_size_mb-1024, 2048))
+    else:
+        mem_limit = str(max(round((config.max_mem_size_mb-1024) / config.cores), 2048))
+
+    mode = backend
+    if config.type == 'regression':
+        mode += '-regression'
+
+    log.info("Running ML-Plan with backend %s in mode %s and a maximum time of %ss on %s cores with %sMB for the JVM, optimizing %s.", backend, mode, config.max_runtime_seconds, config.cores, config.max_mem_size_mb, metric)
+    log.info("Environment: %s", os.environ)
+
+    predictions_file = os.path.join(output_subdir('mlplan_out', config), 'predictions.csv')
+    statistics_file = os.path.join(output_subdir('mlplan_out', config), 'statistics.json')
+
+
+    cmd_root = "java -jar -Xmx{mem_mb}M {here}/lib/mlplan/mlplan-cli*.jar ".format(here=os.path.dirname(__file__),mem_mb=mem_limit)
+    cmd_params = dict(
+        f='"{}"'.format(train_file),
+        p='"{}"'.format(test_file),
+        t=config.max_runtime_seconds,
+        ncpus=config.cores,
+        l=metric,
+        m=mode,
+        s=config.seed,   # weka accepts only int16 as seeds
+        ooab=predictions_file,
+        os=statistics_file,
+        tmp=os.TmpDir,
```

Comment:
sorry, I must have been confusing (and confused myself at the same time): `os.TmpDir` is not  in core `os` module, but in `utils.os` module...
And this is a context manager: compared with the core `tempfile` module, this context cleans after itself, which is mainly useful when running frameworks generating gazillions of artifacts locally:
```python
with utils.os.TmpDir as tmp_dir:
    cmd_params = dict(
        ...
        tmp=tmp_dir,
        **training_params
    )
    ...
    with utils.Timer() as training:
        utils.run_cmd(cmd, _live_output_=True)

```

Reply:
So when I try to use utils.os.TmpDir I obtain the error message that os has no attribute 'TmpDir'. I am not sure whether the system's os module is maybe overriding the one of the utils package...

Reply:
My bad, I should have tried it before mentioning it: `utils` doesn't expose submodules themselves, only their content.
So the right syntax is `utils.TmpDir`.
I tried it locally and seems to work when used as a context manager as above, let me add a suggestion inline.